### PR TITLE
istio: support TLS, TCP traffic routing by virtualservice

### PR DESCRIPTION
**Pull request by tomerlange**
This PR adds edge from virtualservice to pods that TLS, TCP traffic is being routed to by this virtualservice.

### CODE REVIEW

**Comment by tomerlange** [comment]
run skydive-k8s-tests

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-elasticsearch

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

File: topology/probes/istio/virtualservice.go [hunk 1]
```diff
@@ -48,28 +48,31 @@ func newVirtualServiceProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
...
+	TLS  []rule `mapstructure:"tls"`
+	TCP  []rule `mapstructure:"tcp"`
 }
 
-func (vsSpec virtualServiceSpec) getAppsVersions() map[string][]string {
-	appsVersions := make(map[string][]string)
-	for _, http := range vsSpec.HTTP {
-		for _, route := range http.Route {
+func getInstances(rules []rule, instances map[string][]string) map[string][]string {
+	for _, r := range rules {
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: s/r/rule/ as its' `r` is overloaded in this context (could mean rule or route)

File: topology/probes/istio/virtualservice.go [hunk 2]
```diff
@@ -48,28 +48,31 @@ func newVirtualServiceProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
...
+		for _, route := range r.Routes {
 			app := route.Destination.App
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: do we not want to ensure `app != ""`?

File: topology/probes/istio/virtualservice.go [hunk 3]
```diff
@@ -48,28 +48,31 @@ func newVirtualServiceProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
...
 			version := route.Destination.Version
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: do we not want to ensure `version != ""`?

File: topology/probes/istio/virtualservice.go [hunk 4]
```diff
@@ -48,28 +48,31 @@ func newVirtualServiceProbe(client interface{}, g *graph.Graph) k8s.Subprobe {
...
+	HTTP []rule `mapstructure:"http"`
+	TLS  []rule `mapstructure:"tls"`
+	TCP  []rule `mapstructure:"tcp"`
 }
 
-func (vsSpec virtualServiceSpec) getAppsVersions() map[string][]string {
-	appsVersions := make(map[string][]string)
-	for _, http := range vsSpec.HTTP {
-		for _, route := range http.Route {
+func getInstances(rules []rule, instances map[string][]string) map[string][]string {
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: either:

```
func getInstances(rules []rule) map[string][]string {
        ...
}
```

Or

```
func appendInstances(instances map[string][]string, rules []rule) {
        ...
}
```

File: topology/probes/istio/virtualservice.go [hunk 5]
```diff
@@ -82,7 +85,7 @@ func virtualServicePodMetadata(a, b interface{}, typeA, typeB, manager string) g
 	m := k8s.NewEdgeMetadata(manager, typeA)
 	weight := 0
 	for _, http := range vsSpec.HTTP {
-		for _, route := range http.Route {
+		for _, route := range http.Routes {
 			app := route.Destination.App
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: just use `route.Destination.App` directly

File: topology/probes/istio/virtualservice.go [hunk 6]
```diff
@@ -82,7 +85,7 @@ func virtualServicePodMetadata(a, b interface{}, typeA, typeB, manager string) g
...
 			version := route.Destination.Version
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: redundant - just use `route.Destination.Version` directly

File: topology/probes/istio/virtualservice.go [hunk 7]
```diff
@@ -82,7 +85,7 @@ func virtualServicePodMetadata(a, b interface{}, typeA, typeB, manager string) g
...
 			if app == pod.Labels["app"] && version == pod.Labels["version"] {
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: replaced to:

```
if pod.Labels["app"] != "" && pod.Labels["app"] == route.Destination.App &&
                pod.Labels["version"] != "" && pod.Labels["version"] == route.Destination.Version {
         weight += route.Weight
}

File: topology/probes/istio/virtualservice.go (hunk 1 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

File: topology/probes/istio/virtualservice.go (hunk 2 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: This field is mandatory. Error checking will be added in another PR

File: topology/probes/istio/virtualservice.go (hunk 3 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: No. empty version means routing to all versions.

File: topology/probes/istio/virtualservice.go (hunk 4 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

File: topology/probes/istio/virtualservice.go (hunk 5 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

File: topology/probes/istio/virtualservice.go (hunk 7 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

File: topology/probes/istio/virtualservice.go [hunk 8]
```diff
@@ -82,10 +87,8 @@ func virtualServicePodMetadata(a, b interface{}, typeA, typeB, manager string) g
 	m := k8s.NewEdgeMetadata(manager, typeA)
 	weight := 0
 	for _, http := range vsSpec.HTTP {
-		for _, route := range http.Route {
-			app := route.Destination.App
-			version := route.Destination.Version
-			if app == pod.Labels["app"] && version == pod.Labels["version"] {
+		for _, route := range http.Routes {
+			if pod.Labels["app"] != "" && pod.Labels["version"] != "" && route.Destination.App == pod.Labels["app"] && route.Destination.Version == pod.Labels["version"] {
```
Comment (hunchback) [inline topology/probes/istio/virtualservice.go]: If app is equal and route.Destination.Version == "" then you should add the weight

Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

File: topology/probes/istio/virtualservice.go (hunk 6 above)
Comment (tomerlange) [reply topology/probes/istio/virtualservice.go]: DONE

**Comment by tomerlange** [comment]
run skydive-functional-tests-backend-orientdb

**Comment by hunchback** [comment]
@tomerlange - I don't see any tests for this - at minimum please explain how to test this feature - even better add an automatic test.

**Comment by tomerlange** [comment]
I don't think that additional test is necessary as you can check this feature using the existing ones.
1. Edges - can be checked with `TestIstioVirtualServicePodScenario` and `TestBookInfoScenario` - checking specifically the other protocols will be just the same.
2. Weights - relevant only to http traffic, no change here (can be shown in the UI by executing `kubectl create -f tests/istio/virtualservice-pod.yaml`).

**Pull request merged**